src/server/utils/bucket_diff.js

```diff
@@ -235,6 +268,25 @@ class BucketDiff {
         return ans;
     }
 
+    async get_keys_del(first_bucket_contents_left, first_bucket_delete_markers, second_bucket_delete_markers) {
+        const delete_keys = {};
+        const field = 'IsLatest';
+        if (first_bucket_delete_markers.length === 0) {
+            return delete_keys;
+        }
+
+        Object.keys(first_bucket_delete_markers).forEach(key => {
+            const first_bucket_delete_marker = first_bucket_delete_markers[key][field];
+            const second_bucket_delete_marker = second_bucket_delete_markers[key] && second_bucket_delete_markers[key][field];
+            // If the delete marker for a key in the first bucket is latest but is either not present or not latest in the second bucket
+            if (first_bucket_delete_marker && !second_bucket_delete_marker) {
+                delete_keys[key] = first_bucket_contents_left[key];
+            }
+        });
+
+        return delete_keys;
+    }
```

Comment:
I think this logic belongs to the `get_keys_diff` function.
the `get_keys_diff` takes into account the possibility that the list of the first and the secound bucket are not alligned.